components/script/dom/document.rs

```diff
@@ -2014,6 +2019,132 @@ impl Document {
                            ReflowQueryType::NoQuery,
                            ReflowReason::ElementStateChanged);
     }
+
+    // https://fullscreen.spec.whatwg.org/#dom-element-requestfullscreen
+    #[allow(unrooted_must_root)]
+    pub fn enter_fullscreen(&self, pending: &Element) -> Rc<Promise> {
+        // Step 1
+        let promise = Promise::new(self.global().r());
+        let mut error = false;
+
+        // Step 4
+        // check namespace
+        let namespace = pending.namespace().clone();
+        if let Some(dom_str) = Node::namespace_to_string(namespace) {
+            match dom_str.as_ref() {
+                "http://www.w3.org/1999/xhtml" => (),
+                "http://www.w3.org/2000/svg" => {
+                    if pending.local_name().as_ref() != "svg" {
+                        error = true;
+                    }
+                },
+                "http://www.w3.org/1998/Math/MathML" => {
+                    if pending.local_name().as_ref() != "math" {
+                        error = true;
+                    }
+                },
+                _ => {
+                    error = true;
+                }
+            }
+        } else {
+            error = true;
+        }
+        // fullscreen element ready check
+        if !pending.is_connected() {
+            error = true;
+        }
+        // TODO allowfullscreen
+        match self.browsing_context() {
+            None => { error = true; },
+            Some(_) => (),
+        }
+        // TODO fullscreen is supported
+        // TODO This algorithm is allowed to request fullscreen.
+
+        // Step 5
+        let trusted_doc = Trusted::new(self);
+        let trusted_pending = Trusted::new(pending);
+        let trusted_promise = TrustedPromise::new(promise.clone());
+        spawn_named("perform async operation during enter fullscreen".to_owned(), move || {
+            Document::perform_async_enter_fullscreen(trusted_doc, trusted_pending, error, trusted_promise)
+        });
+        promise
+    }
+
+    // https://fullscreen.spec.whatwg.org/#dom-element-requestfullscreen
+    fn perform_async_enter_fullscreen(document: Trusted<Document>,
```

Comment:
We need to derive the document from the element instead of using the one from the original invocation.